FIREFLY-1468 Non-celestial coordinate readout.

tgoldina opened the pull request:
[Jira ticket](https://jira.ipac.caltech.edu/browse/FIREFLY-1468)

Build: https://fireflydev.ipac.caltech.edu/firefly-1468-coordreadout/firefly

- Separated readout options for celestial and non-celestial images
- Added units to projection information
- Added handling for non-degree units when displaying pixel scale
- Added convenience methods to separate all images with projection info into celestial and non-celestial
- For non-celestial coordinates the readout is displayed as `<CTYPE1>, <CTYPE2>: val1, val2 <CUNIT1>`

robyww commented on src/firefly/js/visualize/PlotViewUtil.js:
```diff
@@ -99,12 +99,15 @@ export function getPlotGroupIdxById(ref,plotGroupId) {
  */
-export function hasWCSProjection(ref) {
+export function hasWCSProjection(ref, includeNoncelestial=false) {
     if (!ref) return false;
     const projection= ref.projection ?? primePlot(ref)?.projection;
-    return Boolean(projection?.isSpecified() && projection?.isImplemented());
+    const hasProjection = Boolean(projection?.isSpecified() && projection?.isImplemented());
+    return hasProjection && Boolean(projection?.coordSys?.isCelestial() || includeNoncelestial);
```
good idea to add the parameter!

robyww replied:
I don't think you need the `Boolean` in line 110
```js
return hasProjection && projection?.coordSys?.isCelestial() || includeNoncelestial;
```

robyww commented on src/firefly/js/visualize/ui/MouseReadPopoutAll.jsx:
```diff
@@ -82,7 +82,7 @@ function Readout({readout, readoutData, showHealpixPixel=false, radix}){
     const image= readoutType===STANDARD_READOUT;
 
     if (!get(readoutData,'readoutItems')) return <Box style={rS}/>;
```
I like to removes the `get` when I see it. change to-
```js
if (!readoutData?.readoutItems) return <Box style={rS}/>;
```

robyww commented on src/firefly/js/visualize/ui/MouseReadoutOptionPopups.jsx:
```diff
@@ -22,18 +22,24 @@ import { dispatchChangeReadoutPrefs} from '../../visualize/MouseReadoutCntlr.js'
+];
+
+const noncelestialCoordOptions= [
+	{label: 'WCS Coords', value: 'wcsCoords'},
```
I think we want this to by dynamic. so in the case with will say XANG,YANG. That means you will have to pass the wcs label to the dialog, I think

tgoldina replied:
Done.

robyww commented on src/firefly/js/visualize/ui/MouseReadoutUIUtil.js:
```diff
@@ -56,13 +58,33 @@ export function getNonFluxDisplayElements(readoutItems, readoutPref, isHiPS= fal
+        if (isCelestial) {
+            showReadout1PrefChange= () => showMouseReadoutOptionDialog('imageMouseReadout1', readoutPref.imageMouseReadout1, coordOpTitle);
+        } else {
+            showReadout1PrefChange= () => showMouseReadoutOptionDialog('imageMouseNoncelestialReadout1', readoutPref.imageMouseNoncelestialReadout1, coordOpTitle);
```
I think you will have to pass the wcs label here.

tgoldina commented on src/firefly/js/visualize/VisUtil.js:
```diff
@@ -109,12 +109,12 @@ export function getLatDist(lat1,lat2) {
     if (!wpt) return;
     const from = wpt.getCoordSys();
-    if (!to || from===to) return wpt;
+    if (!to || !from.isCelestial() || from===to) return wpt;
```
Updated docs to clarify that coordinate system change only concerns world points with celestial coordinate system, no action for non-celestial.

tgoldina commented:
Hi Trey, I have addressed your comments and fixed the bugs related to missing `hasWCSProjection()` checks and "LINEAR" CTYPE special case. Could you, please, check it out?

tgoldina commented:
- Fixed Firefly crash when non-celestial image is added to WCS Locked images. WebPlot's getInitZoom now performs extra checks for ZoomType.ARCSEC_PER_SCREEN_PIX type. This is relevant when ZoomType.ARCSEC_PER_SCREEN_PIX is set in WebPlotRequest, the case when WCS Lock is on. (Zoom type is set to `TO_WIDTH_HEIGHT` in these cases. This can be changed later, if it's not the best choice.)
- PlotView's `updateScrollToWcsMatch` now checks for non-celestial images and has no effect on them. However zoom with WCS Lock on affects all images, and this still might need to be fixed.
- Fixed readout for `5713010524-u-size-0.2(deg).fits`. Spatial projection info can now be picked up from alternate WCS, if the it's only set there (main WCS specifies no CTYPEs). This check is done in `makeWebPlotData` of WebPlot.
- I ended up using separate preference for celestial and non-celestial plots both spatial readouts. The reason is that UI allows to switch the readout elements for readout1 and readout2, and sharing preference for the second readout would allow to display an unsupported readout element.
